nucliadb/src/nucliadb/search/api/v1/hydrate.py

```diff
@@ -0,0 +1,758 @@
+            resource=paragraph_id.rid,
+        )
+        self._paragraphs[paragraph_id.full()] = hydrated
+        return hydrated
+
+    def add_page_preview(self, paragraph_id: ParagraphId, page: int, image: Image) -> str:
+        field_id = paragraph_id.field_id
+        field = self._fields[field_id.full()]
+        # TODO: implement link previews
+        assert isinstance(field, hydration_models.HydratedFileField), (
+            f"field type '{field_id.type}' has no preview concept"
+        )
+
+        if field.previews is None:
+            field.previews = {}
+
+        preview_id = f"{page}"
+        field.previews[preview_id] = image
+
+        paragraph = self._paragraphs[paragraph_id.full()]
+        assert paragraph.page is not None, "should already be set"
+        paragraph.page.page_preview_ref = preview_id
+
+        return preview_id
+
+    def add_table_page_preview(self, paragraph_id: ParagraphId, page: int, image: Image) -> str:
+        field_id = paragraph_id.field_id
+        field = self._fields[field_id.full()]
+        # TODO: implement link previews
+        assert isinstance(field, hydration_models.HydratedFileField), (
+            f"field type '{field_id.type}' has no preview concept"
+        )
+
+        if field.previews is None:
+            field.previews = {}
+
+        preview_id = f"{page}"
+        field.previews[preview_id] = image
+
+        paragraph = self._paragraphs[paragraph_id.full()]
+        assert paragraph.table is not None, "should already be set"
+        paragraph.table.page_preview_ref = preview_id
+
+        return preview_id
+
+
+class Hydrator:
+    def __init__(self, kbid: str, config: Hydration):
+        self.kbid = kbid
+        self.config = config
+        self.hydrated = HydratedBuilder()
+
+        # cached paragraphs per field
+        self.field_paragraphs: dict[FieldId, ParagraphIndex] = {}
+
+    async def hydrate(self, paragraph_ids: list[str]) -> Hydrated:
+        driver = get_driver()
+        db_resources = {}
+        async with driver.ro_transaction() as txn:
+            for raw_paragraph_id in paragraph_ids:
+                # TODO: remove this HACK skip
+                if raw_paragraph_id.startswith(":"):
+                    continue
+
+                # TODO if we accept user ids, we must validate them
+                paragraph_id = ParagraphId.from_string(raw_paragraph_id)
+                field_id = paragraph_id.field_id
+                rid = paragraph_id.rid
+
+                if rid not in db_resources:
+                    resource = await get_orm_resource(txn, self.kbid, rid, service_name=SERVICE_NAME)
+                    assert resource is not None
+                    db_resources[rid] = resource
+                resource = db_resources[rid]
+
+                if rid not in self.hydrated.resources:
+                    await self._hydrate_resource(resource, rid)
+
+                if field_id.full() not in self.hydrated.fields:
```

Comment:
are fields always hydrated though? Some may only need the paragraph, maybe? Or at least the option to disable it.

Same goes with resources.